### ISSUE TITLE
Slower binaries since some days

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#42935]
This could have the same  causes of Issue 42808 (i.e. regression due to #42313), or perhaps not.

This small test program solves the Euler problem number 214:

```
#![feature(step_by)]

fn main() {
    fn e214() -> u64 {
        const N_MAX: usize = 40_000_000;
        const SIZE: usize = N_MAX / 2;
        const LEN: u8 = 25;

        let mut chain_length = vec![2u8; SIZE];
        let mut tot: u64 = 0;

        for i in 1 .. SIZE {
            // if not prime, then try next one
            if chain_length[i] > 2 { continue; }

            // p is prime
            let p = 2 * i + 1;

            // calculate q*2^k = p-1
            let mut q = i;
            let mut k = 1;
            while (q & 1) == 0 {
                q /= 2;
                k += 1;
            }

            // calculate cl = chain_length(p)
            let cl = chain_length[q / 2] + k;

            // add p to tot if the length matches.
            if cl == LEN { tot += p as u64; }

            // loop over powers of p.
            let mut pp = p as u64;
            while pp < N_MAX as u64 {
                // Loop over multiples of pp.
                for j in ((pp / 2) .. SIZE as u64).step_by(pp) {
                    chain_length[j as usize] += cl - 2;
                }
                pp *= p as u64;
            }
        }
        tot
    }
    assert_eq!(e214(), 1_677_366_278_943);
}
```

On my PC using nightly-2017-06-18-x86_64-pc-windows-gnu it runs in about 0.39 seconds, while using a more recent Nightly, like the current rustc 1.19.0-nightly (78d8416ca 2017-06-17), it runs in about 0.87 seconds.

**Comment by leonardo-m** [comment]
I compile with:
`rustc -C opt-level=3 -C target-cpu=native -C panic=abort`

**Comment by leonardo-m** [comment]
The old asm:

```
_ZN5test44main17hde37e0367608b3d5E:
	pushq	%r15
	pushq	%r14
	pushq	%rsi
	pushq	%rdi
	pushq	%rbx
	subq	$144, %rsp
	movl	$20000000, %ecx
	movl	$1, %edx
	callq	__rust_allocate
	movq	%rax, %r15
	testq	%r15, %r15
	je	.LBB1_15
	movl	$2, %edx
	movl	$20000000, %r8d
	movq	%r15, %rcx
	callq	memset
	xorl	%r14d, %r14d
	movl	$1, %eax
	jmp	.LBB1_2
	.p2align	4, 0x90
.LBB1_8:
	movb	(%r15,%rdx), %r9b
	movl	%r9d, %edx
	addb	$-25, %dl
	cmpb	%bl, %dl
	movl	$0, %edx
	cmoveq	%rcx, %rdx
	addq	%rdx, %r14
	cmpq	$39999999, %rcx
	ja	.LBB1_2
	leaq	-1(%r8,%r8), %r8
	negb	%bl
	.p2align	4, 0x90
.LBB1_10:
	movq	%rcx, %rdi
	shrq	%rdi
	.p2align	4, 0x90
.LBB1_11:
	movq	%rdi, %rdx
	addq	%rcx, %rdx
	movl	$20000000, %esi
	cmovaeq	%rdx, %rsi
	movzbl	(%r15,%rdi), %edx
	addb	%r9b, %dl
	addb	%bl, %dl
	addb	$-2, %dl
	movb	%dl, (%r15,%rdi)
	movq	%rsi, %rdi
	cmpq	$20000000, %rsi
	jb	.LBB1_11
	movq	%r8, %rdx
	imulq	%rcx, %rdx
	imulq	%r10, %rcx
	leaq	(%rdx,%rcx,2), %rcx
	cmpq	$40000000, %rcx
	jb	.LBB1_10
	.p2align	4, 0x90
.LBB1_2:
	movq	%rax, %r8
	leaq	-1(%r8,%r8), %rcx
	xorl	%r10d, %r10d
	.p2align	4, 0x90
.LBB1_3:
	cmpq	$19999999, %rax
	ja	.LBB1_13
	addq	$1, %r10
	addq	$2, %rcx
	cmpb	$2, (%r15,%rax)
	leaq	1(%rax), %rax
	ja	.LBB1_3
	leaq	-1(%rax), %rdx
	xorl	%ebx, %ebx
	.p2align	4, 0x90
.LBB1_6:
	movq	%rdx, %rsi
	shrq	%rdx
	addb	$-1, %bl
	testb	$1, %sil
	je	.LBB1_6
	cmpq	$40000000, %rsi
	jb	.LBB1_8
	leaq	panic_bounds_check_loc.2(%rip), %rcx
	movl	$20000000, %r8d
	callq	_ZN4core9panicking18panic_bounds_check17hc1b5ea060cf92e68E
	ud2
.LBB1_13:
	movl	$20000000, %edx
	movl	$1, %r8d
	movq	%r15, %rcx
	callq	__rust_deallocate
	movq	%r14, 40(%rsp)
	leaq	40(%rsp), %rax
	movq	%rax, 48(%rsp)
	leaq	ref.4(%rip), %rax
	movq	%rax, 56(%rsp)
	movabsq	$1677366278943, %rax
	cmpq	%rax, %r14
	jne	.LBB1_17
	addq	$144, %rsp
	popq	%rbx
	popq	%rdi
	popq	%rsi
	popq	%r14
	popq	%r15
	retq
.LBB1_15:
	callq	_ZN5alloc3oom3oom17h3058f2f8bf84e7b0E
	ud2
.LBB1_17:
	leaq	48(%rsp), %rax
	movq	%rax, 64(%rsp)
	leaq	_ZN53_$LT$$RF$$u27$a$u20$T$u20$as$u20$core..fmt..Debug$GT$3fmt17h5359b8d7ffbfc5a0E(%rip), %rax
	movq	%rax, 72(%rsp)
	leaq	56(%rsp), %rcx
	movq	%rcx, 80(%rsp)
	movq	%rax, 88(%rsp)
	leaq	ref.9(%rip), %rax
	movq	%rax, 96(%rsp)
	movq	$3, 104(%rsp)
	movq	$0, 112(%rsp)
	leaq	64(%rsp), %rax
	movq	%rax, 128(%rsp)
	movq	$2, 136(%rsp)
	leaq	_ZN5test44main10_FILE_LINE17ha4ac6e59b8b9993dE(%rip), %rdx
	leaq	96(%rsp), %rcx
	callq	_ZN3std9panicking15begin_panic_fmt17h86c214ca1e42d813E
	ud2
```

The new asm:

```
_ZN5test44main17hde37e0367608b3d5E:
	pushq	%r15
	pushq	%r14
	pushq	%r13
	pushq	%r12
	pushq	%rsi
	pushq	%rdi
	pushq	%rbp
	pushq	%rbx
	subq	$152, %rsp
	leaq	64(%rsp), %rcx
	movl	$20000000, %edx
	movl	$1, %r8d
	callq	_ZN5alloc9allocator6Layout15from_size_align17he4ea7424a3dbcbdcE
	cmpq	$0, 64(%rsp)
	je	.LBB1_2
	vmovups	72(%rsp), %xmm0
	vmovaps	%xmm0, 32(%rsp)
	leaq	64(%rsp), %rcx
	leaq	112(%rsp), %r15
	leaq	32(%rsp), %r14
	movq	%r15, %rdx
	movq	%r14, %r8
	callq	_ZN66_$LT$alloc..heap..HeapAlloc$u20$as$u20$alloc..allocator..Alloc$GT$5alloc17h177ce93396a87701E
	cmpq	$0, 64(%rsp)
	jne	.LBB1_8
	movq	72(%rsp), %rbp
	movl	$20000000, %r13d
	movl	$2, %edx
	movl	$20000000, %r8d
	movq	%rbp, %rcx
	callq	memset
	movq	%rbp, 32(%rsp)
	vmovaps	.LCPI1_0(%rip), %xmm0
	vmovups	%xmm0, 40(%rsp)
	xorl	%r12d, %r12d
	movl	$1, %r11d
	movl	$20000000, %r8d
	jmp	.LBB1_15
.LBB1_14:
	movq	32(%rsp), %rbp
	movq	48(%rsp), %r8
	.p2align	4, 0x90
.LBB1_15:
	movq	%r11, %r9
	leaq	-1(%r9,%r9), %rcx
	xorl	%edi, %edi
	movq	%r9, %rdx
	.p2align	4, 0x90
.LBB1_16:
	cmpq	$20000000, %rdx
	jae	.LBB1_17
	cmpq	%rdx, %r8
	jbe	.LBB1_40
	leaq	1(%rdx), %r11
	addq	$1, %rdi
	addq	$2, %rcx
	cmpb	$2, (%rbp,%rdx)
	movq	%r11, %rdx
	ja	.LBB1_16
	leaq	-1(%r11), %rdx
	xorl	%esi, %esi
	.p2align	4, 0x90
.LBB1_29:
	movl	%edx, %ebx
	shrq	%rdx
	addb	$-1, %sil
	testb	$1, %bl
	je	.LBB1_29
	cmpq	%rdx, %r8
	jbe	.LBB1_40
	movb	(%rbp,%rdx), %r10b
	movl	%r10d, %edx
	addb	$-25, %dl
	cmpb	%sil, %dl
	movl	$0, %edx
	cmoveq	%rcx, %rdx
	addq	%rdx, %r12
	cmpq	$39999999, %rcx
	ja	.LBB1_15
	leaq	-1(%r9,%r9), %r9
	negb	%sil
	.p2align	4, 0x90
.LBB1_33:
	movq	%rcx, %rdx
	shrq	%rdx
	.p2align	4, 0x90
.LBB1_34:
	movq	%rdx, %rbp
	addq	%rcx, %rbp
	setb	%bl
	movq	48(%rsp), %r8
	cmpq	%rdx, %r8
	jbe	.LBB1_41
	movq	32(%rsp), %rax
	testb	%bl, %bl
	cmovneq	%r13, %rbp
	movzbl	(%rax,%rdx), %ebx
	addb	%r10b, %bl
	addb	%sil, %bl
	addb	$-2, %bl
	movb	%bl, (%rax,%rdx)
	movq	%rbp, %rdx
	cmpq	$20000000, %rbp
	jb	.LBB1_34
	movq	%r9, %rdx
	imulq	%rcx, %rdx
	imulq	%rdi, %rcx
	leaq	(%rdx,%rcx,2), %rcx
	cmpq	$40000000, %rcx
	jb	.LBB1_33
	jmp	.LBB1_14
.LBB1_17:
	movq	40(%rsp), %rsi
	testq	%rsi, %rsi
	je	.LBB1_38
	leaq	64(%rsp), %rcx
	movl	$1, %edx
	movl	$1, %r8d
	callq	_ZN5alloc9allocator6Layout15from_size_align17he4ea7424a3dbcbdcE
	cmpq	$0, 64(%rsp)
	je	.LBB1_19
	vmovups	72(%rsp), %xmm0
	vmovaps	%xmm0, 112(%rsp)
	leaq	64(%rsp), %rcx
	leaq	112(%rsp), %rdx
	movq	%rsi, %r8
	callq	_ZN5alloc9allocator6Layout6repeat17h9ab668c62bc63d7eE
	cmpq	$0, 64(%rsp)
	je	.LBB1_23
	vmovups	72(%rsp), %xmm0
	leaq	48(%rsp), %rcx
	vmovaps	%xmm0, 64(%rsp)
	leaq	64(%rsp), %r8
	movq	%rbp, %rdx
	callq	_ZN66_$LT$alloc..heap..HeapAlloc$u20$as$u20$alloc..allocator..Alloc$GT$7dealloc17h25a3b92034540e3dE
.LBB1_38:
	movq	%r12, 136(%rsp)
	leaq	136(%rsp), %rax
	movq	%rax, 144(%rsp)
	leaq	ref.4(%rip), %rax
	movq	%rax, 112(%rsp)
	movabsq	$1677366278943, %rax
	cmpq	%rax, %r12
	jne	.LBB1_42
	addq	$152, %rsp
	popq	%rbx
	popq	%rbp
	popq	%rdi
	popq	%rsi
	popq	%r12
	popq	%r13
	popq	%r14
	popq	%r15
	retq
.LBB1_40:
	leaq	panic_bounds_check_loc.2(%rip), %rcx
	callq	_ZN4core9panicking18panic_bounds_check17hc94bb2d7d20ff74cE
	ud2
.LBB1_41:
	leaq	panic_bounds_check_loc.3(%rip), %rcx
	callq	_ZN4core9panicking18panic_bounds_check17hc94bb2d7d20ff74cE
	ud2
.LBB1_2:
.Ltmp12:
	leaq	_ZN38_$LT$core..option..Option$LT$T$GT$$GT$6unwrap14_MSG_FILE_LINE17hc13aed2dd4121e3bE(%rip), %rcx
	callq	_ZN4core9panicking5panic17hc4bc423e36a079b2E
.Ltmp13:
	ud2
.LBB1_8:
	ud2
	ud2
.LBB1_42:
	leaq	144(%rsp), %rax
	movq	%rax, 32(%rsp)
	leaq	_ZN53_$LT$$RF$$u27$a$u20$T$u20$as$u20$core..fmt..Debug$GT$3fmt17ha6ed145788af8ef5E(%rip), %rax
	movq	%rax, 40(%rsp)
	movq	%r15, 48(%rsp)
	movq	%rax, 56(%rsp)
	leaq	ref.9(%rip), %rax
	movq	%rax, 64(%rsp)
	movq	$3, 72(%rsp)
	movq	$0, 80(%rsp)
	movq	%r14, 96(%rsp)
	movq	$2, 104(%rsp)
	leaq	_ZN5test44main10_FILE_LINE17ha4ac6e59b8b9993dE(%rip), %rdx
	leaq	64(%rsp), %rcx
	callq	_ZN3std9panicking15begin_panic_fmt17hb8cee25661fb7188E
	ud2
.LBB1_19:
	leaq	_ZN38_$LT$core..option..Option$LT$T$GT$$GT$6unwrap14_MSG_FILE_LINE17hc13aed2dd4121e3bE(%rip), %rcx
	callq	_ZN4core9panicking5panic17hc4bc423e36a079b2E
	ud2
.LBB1_23:
	leaq	_ZN38_$LT$core..option..Option$LT$T$GT$$GT$6unwrap14_MSG_FILE_LINE17hc13aed2dd4121e3bE(%rip), %rcx
	callq	_ZN4core9panicking5panic17hc4bc423e36a079b2E
	ud2
.LBB1_25:
	jmp	.LBB1_10
.LBB1_21:
	jmp	.LBB1_10
.LBB1_4:
	movq	%rax, %rcx
	callq	rust_eh_unwind_resume
	ud2
.LBB1_11:
	jmp	.LBB1_10
.LBB1_12:
	jmp	.LBB1_10
.LBB1_9:
.LBB1_10:
	movq	%rax, %rcx
	callq	rust_eh_unwind_resume
	ud2
```

**Comment by pnkfelix** [comment]
I also observe a time slowdown on when compiling this benchmark on a Linux system, comparing e00c0401b8 to 1143eb26a2 


e00c0401b8      | 1143eb26a2
---------------- | --------------
real    0m0.265s | real    0m0.690s

**Comment by pnkfelix** [comment]
@leonardo-m it might be good to move those asm dumps to a gist; they're pretty long to scroll through on the thread...

**Comment by pnkfelix** [comment]
(Seems obvious from the asm dumps that one piece of low hanging fruit would be to add some inlining directives to `from_size_align`, I think...)

**Comment by leonardo-m** [comment]
On request I have several other similar small programs that got slower in this timeframe (but their slowdown is smaller. On average I've seen a 10%+ increase of run-time).

> I also observe a time slowdown on when compiling this benchmark

Yes, compilation time of my programs is also increased (perhaps because the compiler as well is slower).

**Comment by pnkfelix** [comment]
> Yes, compilation time of my programs is also increased (perhaps because the compiler as well is slower).

Sorry, I meant "executing this benchmark when compiled with new Allocator support"

**Comment by Mark-Simulacrum** [comment]
Yeah, this was picked up by perf.rlo as well. MIR end regions and then allocators slowed down compilation time in all crates, I believe.

**Comment by pnkfelix** [comment]
(yeah a little embarrassing to have authored both of those PR and have to deal with the fallout now...)

**Comment by pnkfelix** [comment]
A collection of judiciously placed `#[inline]` attributes seems to completely fix the speed regression here.

**Comment by leonardo-m** [comment]
> A collection of judiciously placed #[inline] attributes seems to completely fix the speed regression here.

Very nice. I'll test your changes when they are in a Nightly. And I'll report if there are other performance regressions.

**Comment by pnkfelix** [comment]
However, the `#[inline]` directives make the binary size regression for my microbenchmark on #42808 *worse*

**Comment by pnkfelix** [comment]
Also, it seems like @alexcrichton has independently added many of the same `#[inline]` directives in PR #42727, as he mentioned [here](https://github.com/rust-lang/rust/issues/42808#issuecomment-311502121)

**Comment by leonardo-m** [comment]
Replacing the step_by feature with iterator_step_by, and using .step_by(pp as usize), now the best-of-three run-time is 0.47 seconds.

**Comment by pnkfelix** [comment]
Hmm indeed, seems like the running times regressed further between `(734c83642 2017-07-03)` and `(720c596ec 2017-07-08)` (or maybe I'm not adapting the benchmark correctly...)

**Comment by arielb1** [comment]
1.20 is now beta

**Comment by alexcrichton** [comment]
Looking into this, some things I've learned:

* Technically I don't think this is anything other than a nightly regresion due to the nightly feature usage. Despite this though if you use `RUSTC_BOOTSTRAP=1` to test 1.19.0, 1.20.0, and 1.21.0, you can see that performance regressed in 1.20.0 and was fixed in 1.21.0.
* Historically this program used [`Range::step_by`](https://doc.rust-lang.org/1.19.0/std/ops/struct.Range.html#method.step_by).
* The [`Range::step_by`] function was removed in https://github.com/rust-lang/rust/pull/43012 (not in 1.20.0)
* The new method this program is using is [`Iterator::step_by`](https://github.com/rust-lang/rust/blob/0269acbb10bc7a5b083a5daf095ff391e21ee4e9/src/libcore/iter/mod.rs#L561), introduced in https://github.com/rust-lang/rust/pull/41439 (not in 1.20.0)
* The `StepBy` iterator implementation [calls the `nth` method](https://github.com/rust-lang/rust/blob/0269acbb10bc7a5b083a5daf095ff391e21ee4e9/src/libcore/iter/mod.rs#L561) to do iteration.
* The [`nth` implementation for `ops::Range`](https://github.com/rust-lang/rust/blob/0269acbb10bc7a5b083a5daf095ff391e21ee4e9/src/libcore/iter/range.rs#L240-L251) calls the `add_usize` method
* The [`add_usize` method calls `TryFrom`](https://github.com/rust-lang/rust/blob/0269acbb10bc7a5b083a5daf095ff391e21ee4e9/src/libcore/iter/range.rs#L91-L97)
* On 1.20.0 the number 2 function in the profile is a `TryFrom`-related method
* This method, which should be inlined, was inlined in https://github.com/rust-lang/rust/pull/43248

tl;dr this is fixed, no need for a backport, closing.

**Comment by leonardo-m** [comment]
Thank you, this specific program is indeed fixed (run-time is now 0.40 s), but other performance regressions are not yet fixed, so I'll open other bug reports.

**Comment by alexcrichton** [comment]
Yes that'd be much appreciated, thanks @leonardo-m!
